keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -891,38 +891,65 @@ xref:
+| version skew: ephemeralContainers case           | New pod field, need to be able to validate in same was containers and initContainers if field exists and is populated |
+| version skew: Migration from annotation to field | Need to be able to validate annotation (if present) or field (if it exists and is populated)                          |
+| CRD is deleted                                   | Nothing to type check against, but also means there are no coresponding custom resources                              |
+| CRD is in multiple clusters, but schema differs  | If policy author is aware of the schema variations, can they write policies that work for all the variations?         |
+| Validation of an aggregated API server type      | Main API server does not have type definitions                                                                        |
+
+Due to these complications, Type checking will be optional which can be activated by an expression-level
+`TypeChecking` field. Possible modes are:
+
+- `Never` Do not type check the referred objects or the params any time.
+  Treat `object`, `oldObject`, and `params` as dynamic-typed unstructured objects.
+  This is the default and the current behavior of Kubernetes 1.26.
+- `Informative` type check the referred `object`, `oldObject`, and `params` when the expression is being compiled.
+  Add any errors to the `status` of the given policy. Discard any type information during evaluation and treat referred objects and params unstructured. 
+- `Evaluation` Type check the referred `object`, `oldObject`, and `params` when the expression is being compiled and evaluated after the policy is created. 
+  If the type check fails, mark the policy as misconfigured. Do not type check when the policy is being created or updated.
+- `Always` Type check the referred `object`, `oldObject`, and `params` immediately when the policy is being created or updated. 
+  If the type check fails, including the situation where the schema fails to resolve, the API server will reject the operation on the policy object. 
+  The same type check will happen during compilation and evaluation.
+
+The modes can be summarized as follows.
+
+| Mode        | Value & Param Type | Warning status | Set Policy as misconfigured | Reject on creation/update |
+|-------------|--------------------|----------------|-----------------------------|---------------------------|
+| Never       | Unstructured       | No             | No                          | No                        |
+| Informative | Unstructured       | Yes            | No                          | No                        |
+| Evaluation  | Structured         | Yes            | Yes                         | No                        |
+| Always      | Structured         | Yes            | Yes                         | Yes                       |
```

Comment:
Not blocking, but "Unstructured/Structured" here might be misunderstood to mean that we're using the Kubernetes unstructured type in some cases but the Go structs in others. I'd be okay with omitting the column entirely given that runtime type handling is the same regardless of if we pre-compile or not (https://github.com/kubernetes/enhancements/pull/3779#discussion_r1093562846).

Reply:
`Unstructured` is really too overloaded a term. I am changing them to dynamic vs defined.